src/textual/devtools.py

```diff
@@ -0,0 +1,265 @@
+            console.print(log_message)
+        elif type == "client_spillover":
+            spillover = int(message_json["payload"]["spillover"])
+            info_renderable = DevtoolsInternalMessage(
+                f"Discarded {spillover} messages", level="warning"
+            )
+            console.print(info_renderable)
+        incoming_queue.task_done()
+
+
+async def _consume_outgoing(
+    outgoing_queue: Queue[dict], websocket: WebSocketResponse
+) -> None:
+    while True:
+        message_json = await outgoing_queue.get()
+        type = message_json["type"]
+        if type == "server_info":
+            await websocket.send_json(message_json)
+
+
+async def websocket_handler(request: Request):
+    """aiohttp websocket handler for sending data between devtools client and server"""
+
+    websocket = WebSocketResponse()
+    await websocket.prepare(request)
+    request.app["websockets"].add(websocket)
+
+    console = request.app["console"]
+    size_change_poll_delay = request.app["size_change_poll_delay_secs"]
+
+    incoming_queue: Queue[dict] = Queue()
+    outgoing_queue: Queue[dict] = Queue()
+
+    request.app["tasks"].extend(
+        (
+            asyncio.create_task(_consume_outgoing(outgoing_queue, websocket)),
+            asyncio.create_task(
+                _enqueue_size_changes(
+                    console, outgoing_queue, poll_delay=size_change_poll_delay
+                )
+            ),
+            asyncio.create_task(_consume_incoming(console, incoming_queue)),
+        )
+    )
+
+    console.print(
+        DevtoolsInternalMessage(f"Client '{escape(request.remote)}' connected")
+    )
+
+    await _enqueue_server_info(
+        outgoing_queue, width=console.width, height=console.height
+    )
+    try:
+        async for message in websocket:
+            message = cast(WSMessage, message)
+            if message.type == WSMsgType.TEXT:
+                try:
+                    message_json = json.loads(message.data)
+                except JSONDecodeError:
+                    console.print(escape(str(message.data)))
+                    continue
+
+                type = message_json.get("type")
+                if not type:
+                    continue
+                if type in QUEUEABLE_TYPES:
+                    await incoming_queue.put(message_json)
+            elif message.type == WSMsgType.ERROR:
+                console.print(
+                    DevtoolsInternalMessage("Websocket error occurred", level="error")
+                )
+                break
+    except Exception as error:
+        console.print(DevtoolsInternalMessage(str(error), level="error"))
+    finally:
+        request.app["websockets"].discard(websocket)
+        console.print()
+        console.print(
+            DevtoolsInternalMessage(f"Client '{escape(request.remote)}' disconnected")
+        )
+
+    return websocket
+
+
+async def _on_shutdown(app: Application) -> None:
+    for task in app["tasks"]:
```

Comment:
This is kind of the nuclear option ;-)

It's generally better to ask the tasks to shut down and then wait for them.

For the case of a loop with a websocket, you can close the websocket and the loop will end.

For the cases where you have a `while True` consider making that an [event](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event). i.e. have a single `shutdown_event` and in your tasks do `while not app.shutdown_event.is_set()`. To close all the tasks you can do `self.shutdown_event.set()`

Reply:
Hadn't heard of asyncio Events before, thanks for pointing me in that direction.

Can I check my understanding of why you're describing this as "nuclear" though?

**cancelling tasks:** When I do `task.cancel()`, I'm basically saying I'm happy that a `CancelledError` can be raised at any `await`, and therefore the coroutine could exit at any `await`. My coroutine would have to be designed to be resilient to that possibility.

**events**: If I use an `Event` and set that, I can control _exactly_ where the coroutine will exit. The coroutine will be able to check this event, and ensure that the task ends exactly where I want it to, and we don't need to worry that any `await` could be the exit-point?

Does that sound right?

Reply:
Couldn’t have put it better myself 👍